DEV-1389 Add optional free space check

spacid opened the pull request:
Add an optional check, that checks if there is enough free space on the
target server before transferring the file.

Also, refactor the paramiko client as a instance variable, so that the
client can be reused. This also reduces code copy.

spacid committed:
DEV-1389 Add optional free space check

Add an optional check, that checks if there is enough free space on the
target server before transferring the file.

Also, refactor the paramiko client as a instance variable, so that the
client can be reused. This also reduces code copy.

ghost commented:
## DeepCode's analysis on [#5054ae](https://www.deepcode.ai/app/gh/viaacode/s3-transfer-service/6490c7d1c6f0b88f73a697a15ed25fc409da1206/viaacode/s3-transfer-service/5054ae36b05cc17123bd51c6eae05f2fc418fcaa/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=1&) found:

- :information_source: **1** minor issue. :point_down:
- :heavy_check_mark: **1** issue was fixed.
## Top issues

<table>
  <thead>
    <tr>
      <th align="left">Description</th>
      <th align="left">Example fixes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td width="77%">This self assignment has no impact, consider removing it. Occurrences: <ul><li><a href="https://github.com/viaacode/s3-transfer-service/blob/5054ae36b05cc17123bd51c6eae05f2fc418fcaa/tests/helpers/test_transfer.py#L241">test_transfer.py:241</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/viaacode/s3-transfer-service/6490c7d1c6f0b88f73a697a15ed25fc409da1206/viaacode/s3-transfer-service/5054ae36b05cc17123bd51c6eae05f2fc418fcaa/pr/_/%2Ftests%2Fhelpers%2Ftest_transfer.py/python%2Fdc_interfile_project%2FPythonSelfAssignment/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
  </tbody>
</table>


#### 👉 View analysis in [**DeepCode’s Dashboard**](https://www.deepcode.ai/app/gh/viaacode/s3-transfer-service/6490c7d1c6f0b88f73a697a15ed25fc409da1206/viaacode/s3-transfer-service/5054ae36b05cc17123bd51c6eae05f2fc418fcaa/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=1&) | [_Configure the bot_](https://www.deepcode.ai/app/gh/?ownerconfig=viaacode)
#### 👉 The DeepCode service and API will be deprecated in August, 2021. [Here](https://www.deepcode.ai/blog/migrate-to-snyk-code) is the information how to migrate. Thank you for using DeepCode 🙏 ❤️ !
If you are using our plugins, you might be interested in their successors: [Snyk's JetBrains plugin](https://snyk.co/udpkq) and [Snyk's VS Code plugin](https://snyk.co/udpkr).

spacid commented:
There is a warm shutdown implemented, in the sense that if you send the TERM signal, the service will wait until the transfer is finished. However, the free space check is blocking, meaning that it will wait indefinitely until the space is freed. In the worst case scenario, this means that the service will never stop. In the deployment environment however, there is timeout parameter. If the service hasn't stopped until then, it will terminate the service.

TODO: Adapt the warm shutdown, that when the service is in a free-space-check loop, it just shut downs, sending a NACK (requeue=True) to RABBITMQ and closing the connection form the client side.

maartends commented on docs/s3-t-s_sequence-diagram.plantuml:
```diff
@@ -14,11 +14,18 @@ sts -> rabbit: Listen to queue
+    opt
+        tra -> tra: Check free space
+        loop While not enough free space
+                tra -> tra: Sleep 120 seconds
+                tra -> tra: Check free space
```
```suggestion
                tra -> tra: Check free space
                tra -> tra: Sleep 120 seconds
```

spacid replied:
In the diagram there is a check first if there is enough free space before this loop:
![image](https://user-images.githubusercontent.com/37577903/131812665-0aab19c9-4c7b-470e-9695-3bf20a0373e9.png)

But perhaps this is not the most clear way to visualize the functionality?

maartends commented on README.md:
```diff
@@ -9,6 +9,8 @@ We don't want any traffic in the service itself, so the file is transferred from
 A transfer will be split up in multiple parts of the same size. Each part will be transferred simultaneously, running in a separate thread. Afterwards
 the file will be assembled into the destination file.
 
+There is an optional free space check in which the remote server needs to have a certain amount of free space (in percentage) before the file is transferred. If that free space is not met, the service will sleep indefinitely until the space is freed. The information needed for the check is defined by the ENV vars: `SSH_FREE_SPACE_PERCENTAGE`, `SSH_FILE_SYSTEM`. The vars are mandatory but the valu ...
```
The value needed for `SSH_FILE_SYSTEM` might need some more explanation (I assume it has to be the mountpoint?).

maartends reviewed (approved):
For now, `_check_free_space` will hang indefinitely until space is freed on the remote server. This also without warning (except the log.info "Free space: [...] Space needed: [...]").
Could be a future improvement.

+1 for README and diagram-updates.

spacid committed:
DEV-1389 Remove the superfluous self-assignment

spacid committed:
DEV-1389 Improve documentation

- Change the UML diagram to be more in line with the actual code flow
- README.md:
    - Make it more clear what the ENV var SSH_FILE_SYSTEM is
    - Add exporting the ENV vars of `env.example` to the local test command
